### DISCUSSION TITLE
2023-06-29 CUE Modules Feedback Session / office hours

### DISCUSSION THREAD

**Discussion by myitcv** [cue-lang/cue#2461] [Community Calls]
As announced on the [2026-06-20 community call](https://github.com/cue-lang/cue/discussions/2441), we are starting fortnightly CUE Modules Feedback Sessions / office hours. 

See this discussion for context on the modules proposal: https://github.com/cue-lang/cue/discussions/2330

The first session will be on Thursday, **29 June, 2023 at 1500 UTC**.

Please reply with any questions/topics you would like to see covered.

### Meeting details

https://meet.google.com/nzf-goba-wyn

This session will also be recorded for those unable to join us live.

### CUE Community shared calendar

[Subscribe to the shared calendar](https://calendar.google.com/calendar/u/0?cid=Y19lNzkxMWQ5OWQ4ZGIyMmU2ZTVjMzhkMTVkNjY2ZTVlNjdiNWE5ODNkZWU4N2JmNTU2NDY3NzI1OGIxYjJhMTFhQGdyb3VwLmNhbGVuZGFyLmdvb2dsZS5jb20).  

### Recording

https://youtu.be/_6c7eAeu-Bg

### Slides/materials

https://docs.google.com/presentation/d/1oj3TBWaw-xdj8lXRqQ9anUoTZ_M1ZwHR8yJEmFNMCw0/edit?usp=sharing

### Auto-generated transcript 🆕 

https://docs.google.com/document/d/15K5tIs2RdDj68gm8vivNjKdHw3iEmoOKvufQpZSq_jg/edit?usp=sharing

**Comment by kajogo777** [comment]
Some points to discuss:
- how to handle nested module version resolution (if 2 modules depend on different versions of the same nested module)
- how to handle module integrity checks: file hash vs canonical CUE hash (ignoring whitespace and all the different ways CUE can be written/parsed)
- how can we validate CUE packages in memory, because load still needs disk (cannot use a memory filesystem)

**Comment by verdverm** [comment]
(duplicated from #2330)

From: https://github.com/cue-lang/proposal/blob/main/designs/modules/2449-modules-storage-model.md#detailed-design | https://github.com/cue-lang/proposal/blob/main/designs/modules/2449-modules-storage-model.md#working-with-private-registries

> The location of modules will be determined by a single setting, configured on the command-line by setting $CUE_REGISTRY, that’s used to prefix all module paths in order to find the contents of a module.

> One of the advantages of using the OCI registry model is that it’s possible to use a private or cloud-specific registry to serve content. We see two main requirements there:  (1) Copying modules from the central registry into a local registry. (2) Uploading private modules to a local registry only.


This seems to imply, if I am fetching dependencies, they all must come from the same registry. By proxy, this means if I want to have one private module, I must then mirror all other dependencies.

Is this the case and intention of the design?

Why is fetching from multiple locations in one command, such as public modules from the public registry, and private modules from a private registry, not part of this design?

**Every major package manager supports this, why is CUE deviating from the standard?**  No rationale is provided for this decision

- https://doc.rust-lang.org/cargo/reference/registries.html#using-an-alternate-registry
- https://docs.npmjs.com/cli/v9/configuring-npm/package-json#git-urls-as-dependencies
- https://cloud.google.com/artifact-registry/docs/nodejs#scopes (for how we can scope sets of NPM packages to a registry, GAR in this case)

Most seem to also support multiple registries side-by-side with going directly to source as will, usually limited to git. There is a lot of value in being able to just tag a git commit and have a module available, with no publishing step required.

Given CUE cites these two ecosystems as models, why is the deviation to force a single registry part of the proposal?

**Comment by myitcv** [comment]
Thanks to all those who joined the session. Apologies for the delay, but the description above has been updated with links to the recording, transcript and short slide deck used in the session. 

The second CUE modules feedback session / office hours is happening in a little under an hour (apologies for the late notice). For more details, please see https://github.com/cue-lang/cue/discussions/2487
